### ISSUE TITLE
Guided Trail problem on Android till 30.04!

### ISSUE THREAD

**Issue by iwanmcm** [MathCityMap/O1-MCM-mobile-App#28]
DUE DATE: 30.04
There is an issue with Android (happened only on Samsung devices till now) with the guided trail. 

*UPDATE 19.04*:
How to reproduce:
1) When the route is started the first time the guided trail works without any problem.
2) Inside a trail with active guided mode, go back to the start screen and restart previous trail (which is still then in guided trail mode). After solving the active task, the next task is shortly display and then again dissappears.
My guess: App saves the task number where the user starts from and when solving another task the active task number gets incremented. After restarting the trail this information is somehow lost and when you solve the active task it starts again incrementing from the first start task and displays the second one, although its already active. Maybe it has also something todo with the "back"-button. At this moment it could not be reproduced on iPhone.

Issue:
1) User starts trail.
2) User selects "yes" from "Start From"-Dialog.
3) User selects a task.
4) First task is displayed. User solves first tasks. Second task is displayed. User solves second task and returns to map. Map focuses position of third task but does not display it.

This unfortunatly breaks the padagogical concept behind the guided trail!

Edit: I could reproduce a similiar issue with Huawei, see here:
https://drive.google.com/open?id=1dmrxOga7V-aWBmk_5CptTRWBdfsUg-2d

I was clicking on next task but it did not appear at the first attempt.

What I did before:
1) Start trail and reset all tasks (some task were already solved).
2) Start from task #1 and solved it.
3) Clicked next task, it didnt appear.
4) Go into task details again and click next task, it didnt appear either.
5) On the second attempt it appears.

Please investigate this issue and fix it.

**Comment by iwanmcm** [comment]
Also something with the Close-Symbol is wrong ("X"). It appears twice somehow

**Comment by Mrstolen** [comment]
I looked into this and figured something out:

So, after completing a task and pressing the "next task" button, 2 calls to the redrawMarker() function are made, one when the map initializes (ionViewWillEnter() ) and another form the goToNextTask() function.

For the app to show the correct information, the order of the redrawMarker() calls have to be:
1st the one called from the ionViewWillEnter()
2nd the one called from goToNextTask()

This is what happens in the PC, but on mobile, presumably because of lag or delays, the order is changed, that's why for one split second the information is displayed correctly and right after the map updates to the wrong state.

As a temporary solution, adding a timeout on the goToNextTask() function solves the problem because it is more likely that the ionViewWillEnter() runs first, but I'll try to add a condition that forces the goToNextTask() function to run only after the ionViewWillEnter().

What I haven't figured out yet is if the app should have the updated information already when the ionViewWillEnter() is called.

**Comment by am-immanuel** [comment]
Hi Miguel, thanks for your analysis. A race condition explains, why it does not happen on all devices. I will have a look at the function semantics, then we can discuss a solution. Thanks a lot!

**Comment by Mrstolen** [comment]
The problem was indeed a race condition, but It was not necessary to have two calls to the `redrawMarker` to begin with (I think).

The map was saving it's state when opening a task detail screen, making it so that when returning to the map, the state that he got was not the updated one, instead it was loading the previous one. That is why it was necessary to rely on the second `redrawMarker()`.
By saving the map state on the `goToNextTask()` function and forcing it to run before returning to the map screen, we can have only one `redrawMarker()` (the one called when the map is opened) with the correct updated state.

I haven't tested this 100% but it seems like it's a correct approach to the problem.

**Comment by am-immanuel** [comment]
Perfect solution. Please go ahead!
